### ISSUE TITLE
How to apply credentials (user and password) to Subscriber?

### ISSUE THREAD

**Issue by KrzysztofKielce** [nats-io/stan.net#131]
TL;DR
It would be useful if this example contained information about credentials (username & password):
`https://github.com/nats-io/stan.net/blob/master/examples/stan-sub/stan-sub.cs`

I get exception when I try this code:

	StanOptions stanOptions = StanOptions.GetDefaultOptions();
	stanOptions.NatsURL = _config.nats_URL;
	stanOptions.ConnectTimeout = 5000;

	StanConnectionFactory stanConnectionFactory = new StanConnectionFactory();
	stanConnection = stanConnectionFactory.CreateConnection("my-cluster", "clientID", stanOptions); // exception on this line 
	if (!string.IsNullOrEmpty(_config.nats_username) && !string.IsNullOrEmpty(_config.nats_password))
	{
		stanConnection.NATSConnection.Opts.Password = _config.nats_password;
		stanConnection.NATSConnection.Opts.User = _config.nats_username;
	}

Exception details:
![image](https://user-images.githubusercontent.com/1174432/63077542-edc9a600-bf38-11e9-8fb8-3b98405ef9c8.png)

**Comment by ColinSullivan1** [comment]
You'll want to reorganize your code to set the user/pass in the NATS opts, then set the NATS options in the `stanOptions`.  `natsConn` is included for clarity.  You can simplify a bit by removing the natsConn and just assigning the new connection in place.

```csharp
// Establish a core NATS connection with credentials
var natsOptions = ConnectionFactory.GetDefaultOptions();
natsOptions.Url = _config.nats_URL_rosterservice;
if (!string.IsNullOrEmpty(_config.nats_username) && !string.IsNullOrEmpty(_config.nats_password))
{
    natsOptions.User = _config.nats_username;
    natsOptions.Password = _config.nats_password; ;
}
var natsConn = new ConnectionFactory().CreateConnection(natsOptions);

// Create and assign the stanOptions a new connection
StanOptions stanOptions = StanOptions.GetDefaultOptions();
stanOptions.ConnectTimeout = 5000;
stanOptions.NatsConn = natsConn;
stanConnection = stanConnectionFactory.CreateConnection("my-cluster", "clientID", stanOptions);
```

As STAN is a layer above NATS, we need to setup permissions for the STAN library to use core NATS.  You'll want to setup permissions in the NATS server configuration file.

The STREAMING_SERVER_USER is the credentials that the the NATS streaming server uses to connect.

The **NATS_STREAMING_CLIENT** is the user you use connect to the NATS streaming server, that you would reference in the code above, set in `_config.nats_username`.

```authorization {
NATS_STREAMING_CLIENT = {
      publish = {
        allow = ["_STAN.discover.production", "_STAN.discover.production.>",
                 "_STAN.pub.>", "_STAN.sub.>", "_STAN.close.>",
                 "_STAN.unsub.>", "_STAN.subclose.>", "_INBOX.>"]
      }
      subscribe = {
        allow = ["_INBOX.>", "_STAN.acks.>" ]
      }
    }
    
    STREAMING_SERVER_USER = {
      publish = {
        allow = ["_STAN.>", "_INBOX.>", "raft.>"]
      }
      subscribe = {
        allow = ["_STAN.>", "_INBOX.>", "raft.>"]
      }
    }
    
    users = [
      {user: nss, password: password, permissions: $STREAMING_SERVER_USER}
      {user: clientusername, permissions: $NATS_STREAMING_CLIENT}
    ]
```

If you want to lock it down further by publish subject, you can replace `_STAN.pub.>` with `_STAN.pub.*.your_subject_including_tokens.>`

We know this isn't ideal and will be something we'll address in project JetStream.
